Tsumego

bqv opened the pull request:
Puzzle browser and viewer, from OGS's puzzle repository

![Screenshot_2023-09-30-17-36-27-16_ab1359306de43320f9557c797b1c4be5](https://github.com/acristescu/OnlineGo/assets/822863/3b76000b-0640-4525-80af-303becf9b9e6)

![Screenshot_2023-09-30-17-36-34-26_f94167a5367e07577d817cf990f00b16](https://github.com/acristescu/OnlineGo/assets/822863/a2b3ddf1-8ae3-4fcc-af1e-ec3f74df167a)

![Screenshot_2023-09-30-17-37-08-58_f94167a5367e07577d817cf990f00b16](https://github.com/acristescu/OnlineGo/assets/822863/e9c1820d-40eb-469e-8225-ab7900ffe6fd)

![Screenshot_2023-09-30-17-36-54-28_f94167a5367e07577d817cf990f00b16](https://github.com/acristescu/OnlineGo/assets/822863/d10a5bdc-d761-4126-86b5-b05395870d06)

![Screenshot_2023-09-30-17-37-20-57_ab1359306de43320f9557c797b1c4be5](https://github.com/acristescu/OnlineGo/assets/822863/de6e4a9e-3b8e-4655-944c-bf51e39a2988)

![Screenshot_2023-09-30-17-37-28-36_f94167a5367e07577d817cf990f00b16](https://github.com/acristescu/OnlineGo/assets/822863/7733f913-ff5f-452d-9171-88dc3b82556c)

![Screenshot_2023-09-30-17-37-48-17_f94167a5367e07577d817cf990f00b16](https://github.com/acristescu/OnlineGo/assets/822863/7548577e-e9ab-4a01-8d1b-875bb9bd65f9)

This one should be relatively uncontroversial, though may need a minor cleanup, see what you think

bqv commented:
In my other version of this, I had label marks formatted without the whiteStone background, to be less confusing. I think the code has changed since then, so I left it this time, but it might be a good idea

acristescu commented:
@bqv the latest commit doesn't build I'm afraid:

```
> Task :app:kspDebugKotlin FAILED
e: [ksp] /bitrise/src/app/src/main/java/io/zenandroid/onlinego/data/db/GameDao.kt:40: [MissingType]: Element 'io.zenandroid.onlinego.data.db.GameDao' references a type that is not present
e: Error occurred in KSP, check log for detail
```

bqv commented:
My bad, I'm juggling quite a few branches 😅 I'm not entirely done on fixing things though, but I'll sort that

acristescu commented:
No rush, I thought I'd mention it since I don't think you can see the Bitrise build log (right?).

bqv commented:
Could you rerun CI? I'm not sure that even makes sense, the error is in inline sql, and that commit does build here

bqv commented:
I'm confident the issue isn't me, I'm assuming there's some version discrepancy between me and CI that's causing unnatural errors, but we can figure that out later. If you get a sec, I'd appreciate if you could try building it manually, though

bqv commented:
Just these two left if I'm reading right

- [ ] MVI pattern 
- [ ] RxJava instead of flows

I'll chip away at those slowly, unless you're happy to get it in before that

acristescu commented:
OMG that's a lot of PRs all at the same time :) But I remember now, you're that guy that de-google-ified the app all on his own, then published it on a FOSS repository! I discovered your published up a year or so ago and I thought to myself that's a lot of work that could have been used on the main project instead. I planned to backport the features myself but I couldn't find the time really and the tech stack had diverged too much since (as you can see).

Thank you @bqv for taking the time to bring these in to the main app, I realize it's a LOT of work. 

Since this is not new code (and also because there is a LOT of it), perhaps we can bend the rules a bit. For example rewriting everything to use Flow instead of RxJava would be too much work IMHO. There is still quite a bit of RxJava stuff in the app so adding a bit more of it is not the end of the world.

I would still like to see the MVI part removed if possible though. If you think that's too hard, we can consider merging it as it is and then I can slowly chip at it. In particular this one PR is in a bit of a weird spot right now, as it has both a `StateFlow` AND a `render` method. Makes it really confusing.

What's your take on this? In either case, I'd rather we focus on one PR at at time as 8 of them at once is too much.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/local/Puzzle.kt:
```diff
@@ -0,0 +1,62 @@
+
+@Entity
+data class Puzzle(
+    @PrimaryKey var id: Long,
```
All of the fields of the model need to be `val` and not `var`. We should also mark the data classes as `@Immutable`, particularly if the contain `List` objects. This is not just pedantic, it's because of how Compose works: If the state is immutable, then many recompositions can be skipped and performance drastically improves. See Google's recommendations here https://developer.android.com/jetpack/compose/performance/stability

And yes, I know I haven't changed it everywhere myself.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/local/PuzzleCollection.kt:
```diff
@@ -0,0 +1,58 @@
+
+@Entity
+data class PuzzleCollection(
+    @PrimaryKey var id: Long,
```
val instead of var, `@Immutable` (see other comment)

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/local/PuzzleCollectionSolutionMetadata.kt:
```diff
@@ -0,0 +1,9 @@
+
+@Entity
+data class PuzzleCollectionSolutionMetadata constructor(
+    var collectionId: Long,
```
val instead of var, `@Immutable` (see other comment)

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/local/VisitedPuzzleCollection.kt:
```diff
@@ -0,0 +1,19 @@
+@RequiresApi(Build.VERSION_CODES.O)
+@Entity
+data class VisitedPuzzleCollection constructor(
+    var collectionId: Long,
```
val instead of var, `@Immutable` (see other comment)

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/ogs/OGSPlayer.kt:
```diff
@@ -1,24 +1,28 @@
 /**
  * Created by alex on 04/11/2017.
  */
+@Entity
```
Pls don't put OGSPlayer in the database, put Player instead.

bqv replied:
This actually already was removed from the DB, just needed to drop the Entity annotation

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/ogs/OGSPuzzle.kt:
```diff
@@ -0,0 +1,95 @@
+import java.time.Instant
+
+data class MoveTree (
+    var y: Int = -1,
```
val instead of var, `@Immutable` (see other comment)

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/ogs/OGSPuzzleCollection.kt:
```diff
@@ -0,0 +1,35 @@
+import java.time.Instant
+
+data class OGSPuzzleCollection (
+    var id: Long = -1,
```
val instead of var, `@Immutable` (see other comment)

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/model/ogs/PuzzleSolution.kt:
```diff
@@ -0,0 +1,51 @@
+@Entity
+data class PuzzleSolution (
+    @PrimaryKey var id: Long? = null,
+    var puzzle: Long = -1,
```
val instead of var, `@Immutable` (see other comment)

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/repositories/PuzzleRepository.kt:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    private fun onError(error: Throwable) {
+        Log.e("PuzzleRepository", error.message, error)
```
please use `recordException(error)` so that the error is captured by Crashlytics.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/di/Modules.kt:
```diff
@@ -176,6 +180,61 @@ private val viewModelsModule = module {
+  viewModel {
+    PuzzleDirectoryViewModel(
+      PuzzleRepository(get(), get()),
+      OGSRestService(get(), get(), get(), get()),
```
you're specifically instantiating a new instance of `OGSRestService()` here, but I'm trying to keep that a singleton. you should be able to just `get()` here.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/di/Modules.kt:
```diff
@@ -176,6 +180,61 @@ private val viewModelsModule = module {
 
+  viewModel {
+    PuzzleDirectoryViewModel(
+      PuzzleRepository(get(), get()),
```
just `get()` the instance of PuzzleRepository from the DI graph

bqv replied:
I'd spotted that myself too but forgot to do anything about that..

acristescu commented on app/src/main/java/io/zenandroid/onlinego/di/Modules.kt:
```diff
@@ -176,6 +180,61 @@ private val viewModelsModule = module {
+
+  viewModel { params ->
+    PuzzleViewModel(
+      PuzzleRepository(get(), get()),
```
get()

acristescu commented on app/src/main/java/io/zenandroid/onlinego/di/Modules.kt:
```diff
@@ -176,6 +180,61 @@ private val viewModelsModule = module {
+  viewModel { params ->
+    PuzzleViewModel(
+      PuzzleRepository(get(), get()),
+      OGSRestService(get(), get(), get(), get()),
```
get()

acristescu commented on app/src/main/java/io/zenandroid/onlinego/di/Modules.kt:
```diff
@@ -176,6 +180,61 @@ private val viewModelsModule = module {
+
+  viewModel { params ->
+    TsumegoViewModel(
+      PuzzleRepository(get(), get()),
```
get()

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/DirectoryAnalyticsMiddleware.kt:
```diff
@@ -0,0 +1,18 @@
+import io.zenandroid.onlinego.ui.screens.puzzle.PuzzleDirectoryAction.*
+import io.zenandroid.onlinego.mvi.Middleware
+
+class DirectoryAnalyticsMiddleware: Middleware<PuzzleDirectoryState, PuzzleDirectoryAction> {
```
Does this do anything or can we remove it?

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleAnalyticsMiddleware.kt:
```diff
@@ -0,0 +1,18 @@
+import io.zenandroid.onlinego.ui.screens.puzzle.PuzzleAction.*
+import io.zenandroid.onlinego.mvi.Middleware
+
+class PuzzleAnalyticsMiddleware: Middleware<PuzzleState, PuzzleAction> {
```
Maybe remove this?

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryFragment.kt:
Can we move this class (and dependent ones, such as view model) to a different package than the `Tsumego` please? I would also like to keep the naming consistent, such as "PuzzleList" and "Puzzle", in my mind Tsumego and Puzzle are synonims.

bqv replied:
They are synonyms, that was a naming fail on my part, I started from PuzzleDirectory and Puzzle and realised I needed another screen. Perhaps PuzzleDirectory, PuzzleListing, Puzzle?

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryState.kt:
```diff
@@ -0,0 +1,11 @@
+import io.zenandroid.onlinego.data.model.local.PuzzleCollection
+import java.time.Instant
+
+data class PuzzleDirectoryState (
```
Sadly, The way kotlin works with JVM, it cannot guarantee immutability of a Map. We have to add `@Immutable` here.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryViewModel.kt:
```diff
@@ -0,0 +1,92 @@
+        viewBinding?.dispose()
+    }
+
+    private fun onError(t: Throwable) {
```
`recordException(error)` please

bqv replied:
I also removed the toasts, I think I was mainly using them for debugging and they slightly violate the paradigm

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleFragment.kt:
```diff
@@ -0,0 +1,285 @@
+                DisposeOnLifecycleDestroyed(viewLifecycleOwner)
+            )
+            setContent {
+                OnlineGoTheme {
```
in order to ensure that the composable does not mistakenly use any state from the Fragment, I like to bring the Composable out of the fragment scope, sometimes even in a different file (e.g. `PuzzleScreenUI.kt`).

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleFragment.kt:
```diff
@@ -0,0 +1,285 @@
+            )
+            setContent {
+                OnlineGoTheme {
+                    val state by rememberStateWithLifecycle(viewModel.state)
```
state is not null, so all "?" and "!!" are not needed

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleFragment.kt:
```diff
@@ -0,0 +1,285 @@
+                    ) {
+                        val titleState = remember {
+                            val base = "Puzzles"
+                            derivedStateOf {
```
if the data class is immutable (vals, etc) then the `derivedStateOf` is not needed

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleFragment.kt:
```diff
@@ -0,0 +1,285 @@
+                            ) {
+                                items(items = puzzles) {
+                                    val solvedState = remember {
+                                        derivedStateOf {
```
not really needed if the state is immutable

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleFragment.kt:
```diff
@@ -0,0 +1,285 @@
+                                            }
+                                        }
+                                    }
+                                    LaunchedEffect(true) {
```
This should be launched from the ViewModel

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoState.kt:
```diff
@@ -0,0 +1,31 @@
+import io.zenandroid.onlinego.data.model.StoneType
+import java.time.Instant
+
+data class TsumegoState (
```
`@Immutable`

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoViewModel.kt:
```diff
@@ -0,0 +1,329 @@
+    }
+
+    private fun setPuzzle(puzzle: Puzzle, attempts: Int = 1) {
+        _state.update {
```
this might actually be better than the method I was using `_state.value = _state.value.copy(...)` as it's atomic. I will make a mental note to use it myself.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/data/db/Database.kt:
```diff
@@ -6,17 +6,34 @@ import androidx.room.RoomDatabase
+            ChallengeNotification::class,
+            PuzzleCollection::class,
+            Puzzle::class,
+            PuzzleRating::class,
```
why do we need to put the rating in a different table?

bqv replied:
It's a separate API call and object, it just seemed simpler. Are you thinking it's worth merging it into Puzzle?

bqv replied:
We do offer that actually, they can be rated from the Tsumego screen, but I'll see about merging them

bqv replied:
Just a thought - I think the reason I kept those separate was because if the user logs out and logs in as another account, the puzzle table can be reused, but the ratings will all need to be refreshed because they're per-user. Is that enough of a reason to keep them as they are?

acristescu reviewed (changes requested):
Hi @bqv I've done a first pass of the PR and mark some of the issues. Please let me know if you want me to handle some of these issues and how should we do it to avoid big merge conflicts.

bqv commented:
No problem, I can handle all of those! I think I'd check the contributing PR, as that's just one or two lines and is just a QoL thing, but otherwise I agree let's get this somewhere first

bqv commented:
For the record, in some places you have also used javax.annotation.concurrent.Immutable instead of androidx.compose.runtime.Immutable, I'm not sure if this is intentional

acristescu commented:
> For the record, in some places you have also used javax.annotation.concurrent.Immutable instead of androidx.compose.runtime.Immutable, I'm not sure if this is intentional

ah, nice catch, indeed it's not intentional. I'll make the change on my repo.

bqv commented:
I've made the naming less ambiguous, but you're free to rename them if you think there's a better option.

Just two notes left for you on that review, otherwise I'll see about the mvi removal

acristescu commented:
@bqv we're definitely getting there. I had a few minutes to spare on this today so I moved the code around a bit just for the Puzzle Set screen. For example I removed the not needed `object PuzzleSetUI` and removed the injection into the composable (I think that made the entire composable unstable, but don't quote me on that).

Try as I might, I could not open a PR on your repo though (I think it's protected), so I'm attaching the diff here 
[patch.zip](https://github.com/acristescu/OnlineGo/files/12869908/patch.zip)

Can you please have a look, apply it if you're OK with it and do the same for the other 2 screens? I do think the core code is looking great, I'm just trying to have the same pattern for all screens to help with maintenance and also because as soon as I finish nuking all the legacy views I'm planning to refactor the entire repository to not use any fragments at all and use compose navigation instead.

That task would be a lot easier if everything works the same way and we don't have odd screens (as we do right now with the AI screen for example).

bqv commented:
> @bqv we're definitely getting there. I had a few minutes to spare on this today so I moved the code around a bit just for the Puzzle Set screen. For example I removed the not needed `object PuzzleSetUI` and removed the injection into the composable (I think that made the entire composable unstable, but don't quote me on that).
The boardTheme injection I assume you mean, yeah I wasn't sure about that but it felt weird to have that in the state. That's fine

> Try as I might, I could not open a PR on your repo though (I think it's protected), so I'm attaching the diff here [patch.zip](https://github.com/acristescu/OnlineGo/files/12869908/patch.zip)
No problem, `git am`'d, I'm not sure why my repo is protected or how to unprotect it though, but I added you as a collaborator, I assume that'll solve it

> Can you please have a look, apply it if you're OK with it and do the same for the other 2 screens? I do think the core code is looking great, I'm just trying to have the same pattern for all screens to help with maintenance and also because as soon as I finish nuking all the legacy views I'm planning to refactor the entire repository to not use any fragments at all and use compose navigation instead.
> That task would be a lot easier if everything works the same way and we don't have odd screens (as we do right now with the AI screen for example).
Sounds good, done also (I didn't realise the onBackPressedCallback trick was not needed, I guess I just copied that from wherever I got this screen template). Sorry for the messy imports, something about this project makes android studio choke badly so I've always been developing it by hand in a CLI

acristescu commented:
> something about this project makes android studio choke badly so I've always been developing it by hand in a CLI

That is really weird, I've worked on this project on 3 different machines, on both windows and Mac and never had an issue. In fact it's quite a bit lighter than some of the projects I've encountered at work. The only thing I can think of is that I'm using Canary channel Android Studio, maybe you're using something more stable?

Anyway, the structure of the feature looks OK to me now, I'm going to have a review of the Compose stuff and once that's done we can merge it in. Sorry for giving this piecemeal to you, but it's quite a big PR to wrap my head around in a single go.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleSetUI.kt:
```diff
@@ -0,0 +1,161 @@
+            }
+          }
+          LaunchedEffect(true) {
+            fetchSolutions(it.id)
```
Why are we triggering data loading from here? Are we trying to do it when the item appears on the screen? Won't this cause a large number of requests to be triggered if the user scrolls fast?

bqv replied:
Yes! It would, I suppose, but the aim is to avoid launching all the requests at once, because doing that on the regular, for hundreds(?) of users, sounds like a way to upset anoek. Making it work by visibility at least minimizes that, but actually yes I would like a nice solution to filtering out the "scrolls past" requests, is there one?

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoAction.kt:
```diff
@@ -0,0 +1,31 @@
+import io.zenandroid.onlinego.data.model.local.Puzzle
+import io.zenandroid.onlinego.data.model.local.PuzzleCollection
+
+sealed class TsumegoAction {
```
minor: in the latest iteration of the Kotlin language this should be a "sealed Interface" and the actual objects and classes should be "data object" or "data classes". But it doesn't really matter that much.

bqv replied:
no worries, will change

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoUI.kt:
```diff
@@ -0,0 +1,249 @@
+    Column (
+        modifier = Modifier.fillMaxHeight()
+    ) {
+        val titleState = remember {
```
I haven't tried it, but I don't think remember and derivestateOf are needed here.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoUI.kt:
```diff
@@ -0,0 +1,249 @@
+                            boardTheme = state.boardTheme,
+                            drawCoordinates = state.drawCoordinates,
+                            interactive = state.boardInteractive,
+                            drawShadow = false,
```
why no shadow? That feature was added to increase performance for thumbnail boards presented in lists.

bqv replied:
I actually don't remember why I set this. I'll add it back unless I remember why

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoUI.kt:
```diff
@@ -0,0 +1,249 @@
+                shape = MaterialTheme.shapes.medium,
+                modifier = Modifier.fillMaxSize()
+            ) {
+              //val listener = { action: TsumegoAction ->
```
dead code

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/TsumegoUI.kt:
```diff
@@ -0,0 +1,249 @@
+        }
+    }
+    LaunchedEffect(solved) {
+        snapshotFlow { solved }
```
This looks like logic that needs to be offloaded to the ViewModel. Not sure I even understand it, is that a way of saying "if it's solved, call onSolve()"?

bqv replied:
yes, pretty much. It should be in the VM, you're right

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+    val listState = rememberLazyListState()
+    var filterText by remember { mutableStateOf(TextFieldValue()) }
+    var sortField by remember { mutableStateOf<PuzzleDirectorySort>(RatingSort(false)) }
+    val resultCollections = filterText.text.lowercase().let { query ->
```
This has the potential to be run every frame, at the very least it should be remembered, but the proper way to do it is to offload this logic to the ViewModel. The Composable should be concerned only with displaying stuff, not filtering or really remembering much.

bqv replied:
Agreed, ok

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+                        .fillMaxWidth()
+                        .padding(horizontal = 8.dp, vertical = 4.dp)
+                ) {
+                    Row(modifier = Modifier
```
There is too much info and text displayed for a mobile app. Please consider removing most of the stuff that can be removed and still keep functionality (such as the number of views or the number of solves). The whole idea of the app is to provide a streamlined experience to the user, if the user wants all the features he or she can just use the clunky website.

bqv replied:
I'd rather not drive people to using the website if possible, that's what most of my efforts have been in the interest of. I don't mind hiding some of this though, if it's still accessible, somewhere, I just don't all the ideas for that, maybe a popup for further detail? More little flags like the one for "visited"?

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+import kotlin.reflect.full.safeCast
+
+@Composable
+fun PuzzleDirectoryScreen(
```
This screen sadly has a few layout issues on Samsung Galaxy S21 Ultra (the most used family of phones for this app). Here is a screenshot:

![Screenshot_20231012_164144](https://github.com/acristescu/OnlineGo/assets/2835035/52b28ad1-eff7-4e23-be47-82fa8f87c5f5)

bqv replied:
Oh, gods, ok. I might be able to simulate that in waydroid, but help would be appreciated

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+        modifier = Modifier.fillMaxHeight()
+    ) {
+        stickyHeader {
+            TopAppBar(
```
It makes no difference beyond semantics, does it. Sure

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+                        modifier = Modifier.fillMaxWidth().padding(horizontal = 8.dp),
+                    ) {
+                        @Composable
+                        fun <T : PuzzleDirectorySort> ratingButton(type: KClass<T>, name: String) {
```
I ...don't know :D I might have an answer when I look closer at this, but at a glance it seems like it was just an attempt at dedup and I might have got carried away

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+                    }
+
+                    if (filterText.text.isEmpty()) {
+                        LazyRow(modifier = Modifier.fillMaxWidth()) {
```
It's horizontal, because it displays all the visited puzzle sets, but you'll have to horizontally scroll to see ones further back in history - try visiting more than three to see an example. It's lazy so that that's cheap. I don't think it's a problem, the horizontal scroll isn't even accessible unless you know its there, and as I said is free, but up to you if you want it

bqv replied:
There might be something to consider in dropping the gap when there's less then three visited, I agree that's kinda unnecessary and garish

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+                                                                    Text(
+                                                                        text = "by ${it.username} $flag - visited $ago",
+                                                                        style = TextStyle.Default.copy(
+                                                                            fontSize = 8.sp,
```
It's actually a pretty decent size for me, I think you can see from screenshots, is SP a good unit to be using?

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+                                                                        fontWeight = FontWeight.Bold
+                                                                    )
+                                                                )
+                                                                it.owner?.let {
```
Smaller size then? or different unit from SP? Again, looks alright on all the phones I've tried this on (per my screenshots)

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,418 @@
+                Surface(
+                    shape = MaterialTheme.shapes.medium,
+                    modifier = Modifier
+                        .height(150.dp)
```
To be honest, I have no idea what I'm doing frontend-wise, suffice to say I'll leave UI choices to you, I'm not entirely sure what's wrong here

bqv commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleSetUI.kt:
```diff
@@ -0,0 +1,161 @@
+              .padding(horizontal = 10.dp, vertical = 10.dp)) {
+              Row {
+                it.puzzle.let {
+                  val pos = RulesManager.buildPos(
```
would laziness go somewhere in solving that, or is it only going to be solved when in viewmodel - purely out of curiosity

bqv replied:
Looking closer, this only applies to the puzzles on the screen, so maybe 3 or so at most. Done once keyed on state.puzzle, i don't see an issue

bqv commented:
Resolved many of those, but the remaining problems are things I'm not sure how to address. In resolving those I also got rid of some rxjava, rather than adding more, so I might extend that operation to the rest

bqv commented:
Made some changes. Could you try the UI again, see if that satisfies it?

acristescu commented:
> Made some changes. Could you try the UI again, see if that satisfies it?

It looks better, as most things are not cut off now, but some things are still problematic. Chief amongst them is the fact that we're computing the position in Compose which causes the scrolling to stutter. Also the sorting bar loos a bit weird. I can rewrite the whole Compose for that screen I think after we merge. I'll try to see when I have time to do that.

![image](https://github.com/acristescu/OnlineGo/assets/2835035/c29f223b-6981-4bb7-ad67-fdbffb461a09)

acristescu reviewed (approved):
@bqv In the interest of keeping things moving forwards, I'm going to merge this and then do some fixes on the merged version. Please feel free to open small PRs on it if you have further things to add.

acristescu commented on app/src/main/java/io/zenandroid/onlinego/ui/screens/puzzle/PuzzleDirectoryUI.kt:
```diff
@@ -0,0 +1,432 @@
+            state.collections.map { it.value }.ifEmpty { null }?.let { collections ->
+                item {
+                    var compositionCount by remember { mutableStateOf(0) }
+                    SideEffect { compositionCount++ }
```
@bqv FYI: this doesn't work as increasing the compositionCount causes a recomposition, which in turn causes a new recomposition (as you're reading compositionCount in the same scope) which causes the SideEffect to trigger, which causes the recomposition to occur, etc... This is a rather interesting case of backwards write: https://developer.android.com/jetpack/compose/side-effects#sideeffect-publish

bqv replied:
ah! I wondered what was happening. I actually left that in by accident, it was me trying to debug something

bqv commented on app/src/main/java/io/zenandroid/onlinego/data/repositories/PuzzleRepository.kt:
```diff
@@ -35,150 +47,144 @@ class PuzzleRepository(
-                    disposable += restService.getPuzzleCollections()
-                            .subscribe(this::saveCollectionsToDB, this::onError)
+
+        flow<Int> { dao.getPuzzleCollectionCount() }
```
I believe I did, for many, can't remember why this was missed

bqv commented:
No prob, thanks. 


> Is rating the puzzle (or even viewing the rating of a puzzle) really a feature we want? I would say no...

I would say yes, mainly as I use it myself to keep track of puzzles I want to remember are worthwhile.

Also, of course, sorting by rating is about the best way to get the useful ones

Noted though, on the features discussion. I'd still like to have many of those, but I'm happy to keep those in my repo as patches at worst

acristescu commented:
@bqv if you're interested, I've made a commit on master that RxJava completely from the tsumego. I'm now getting HTTP 429s when getting the solves for the puzzles, so I'm considering what to do there. Maybe not get them at all? Maybe skip that entire screen as the website does?

bqv commented:
> I'm now getting HTTP 429s when getting the solves for the puzzles, so I'm considering what to do there. Maybe not get them at all? Maybe skip that entire screen as the website does?

yeah, thats what I was saying before, the only sane way to get them is to lazily fetch them as displayed. I think basically just leave them as they were and have a placeholder image that gets fulfilled on an IO thread, if the lag bothers you?
